### PR TITLE
DynamoDB ConditionalPut

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

Builds on #4918 and extends it to conditional put.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/aws/dynamo.rs [hunk 1]
```diff
@@ -61,16 +63,24 @@ const STORE: &str = "DynamoDB";
 ///
 /// The DynamoDB schema is as follows:
 ///
-/// * A string hash key named `"key"`
+/// * A string partition key named `"path"`
+/// * A string sort key named `"etag"`
```
Comment (tustvold) [inline object_store/src/aws/dynamo.rs:67]: This is a breaking change, but given #4918 was only merged this morning and hasn't been released I think that is fine

File: object_store/src/aws/precondition.rs [hunk 2]
```diff
@@ -48,7 +48,7 @@ pub enum S3CopyIfNotExists {
     HeaderWithStatus(String, String, reqwest::StatusCode),
     /// The name of a DynamoDB table to use for coordination
     ///
-    /// Encoded as either `dynamodb:<TABLE_NAME>` or `dynamodb:<TABLE_NAME>:<TIMEOUT_MILLIS>`
+    /// Encoded as either `dynamo:<TABLE_NAME>` or `dynamo:<TABLE_NAME>:<TIMEOUT_MILLIS>`
```
Comment (tustvold) [inline object_store/src/aws/precondition.rs:51]: Drive by docs fix

File: object_store/src/aws/dynamo.rs [hunk 3]
```diff
@@ -61,16 +63,24 @@ const STORE: &str = "DynamoDB";
...
 ///
-/// To perform a conditional operation on an object with a given `path` and `etag` (if exists),
+/// An appropriate DynamoDB table can be created with the CLI as follows:
+///
+/// ```bash
+/// $ aws dynamodb create-table --table-name <TABLE_NAME> --key-schema AttributeName=path,KeyType=HASH AttributeName=etag,KeyType=RANGE --attribute-definitions AttributeName=path,AttributeType=S AttributeName=etag,AttributeType=S
+/// $ aws dynamodb update-time-to-live --table-name <TABLE_NAME> --time-to-live-specification Enabled=true,AttributeName=ttl
+/// ```
+///
+/// To perform a conditional operation on an object with a given `path` and `etag` (`*` if creating),
```
Comment (tustvold) [inline object_store/src/aws/dynamo.rs:79]: DynamoDB doesn't let you use an empty string for a key attribute, so we use `*`, which is an illegal ETAG value (as If-Match wildcards would then be ambiguous)

**Comment by tustvold** [comment]
@roeap when you get a chance, perhaps you could give this a look over

**Comment by tustvold** [comment]
@roeap sorry to dump this on you, but I just realised this needs to go in the release or it'll be a major breaking change... Could you perhaps take a look? :pray:

**Comment by roeap** [comment]
@tustvold - I'll take a look.

**Review by roeap — APPROVED** [review]
👍 - looking good, excited to see this land.

**Pull request merged**